components/src/preact/components/downshift-combobox.tsx

```diff
@@ -0,0 +1,159 @@
+import { useCombobox } from 'downshift/preact';
+import { useMemo, useRef, useState } from 'preact/hooks';
+import { type JSXInternal } from 'preact/src/jsx';
+
+export function DownshiftCombobox<Item, Data>({
+    allData,
+    convertToItem,
+    value,
+    filterItemsByInputValue,
+    createEvent,
+    itemToString,
+    emptyEvent,
+    placeholderText,
+    formatItemInList,
+}: {
+    allData: Data[];
+    convertToItem: (data: Data) => Item | undefined;
+    value?: Data;
+    filterItemsByInputValue: (item: Item, value: string) => boolean;
+    createEvent: (item: Item) => CustomEvent;
+    itemToString: (item: Item | undefined | null) => string;
+    emptyEvent: () => CustomEvent;
+    placeholderText?: string;
+    formatItemInList: (item: Item) => JSXInternal.Element;
+}) {
+    const allItems = useMemo(
+        () =>
+            allData
+                .map((data) => {
+                    return convertToItem(data);
+                })
+                .filter((item): item is Item => item !== undefined),
+        [allData, convertToItem],
+    );
+
+    const initialSelectedItem = useMemo(
+        () => (value !== undefined ? convertToItem(value) : null),
+        [value, convertToItem],
+    );
+
+    const [items, setItems] = useState(
+        allItems.filter((item) => filterItemsByInputValue(item, itemToString(initialSelectedItem))),
```

Comment:
Doesn't it make more sense to hand the full `initialSelectedItem` to that function? The consuming code can still decide to turn it into a string somehow.
```suggestion
        allItems.filter((item) => filterItemsByInputValue(item, initialSelectedItem)),
```

Reply:
I think it should be like this. The idea is to filter the items based on the input from the input field. So the list changes, when the user types. The itemToString option is here necessary, since the "value" must me converted to a string first.